CommonMark/ExtensibleHtmlPrinter.cs

```diff
@@ -0,0 +1,742 @@
+        {
+            if (_trackPosition)
+            {
+                WriteConstant("<hr");
+                PrintPosition(block);
+                writer.WriteLine();
+            }
+            else
+            {
+                WriteLineConstant("<hr />");
+            }
+        }
+
+        protected virtual void PrintHtmlBlock(TextWriter writer, Block block)
+        {
+            // cannot output source position for HTML blocks
+            block.StringContent.WriteTo(writer);
+        }
+
+        protected virtual void PrintCode(TextWriter writer, Block block)
+        {
+            int x;
+            EnsureLine();
+            WriteConstant("<pre><code");
+            if (_trackPosition) PrintPosition(block);
+
+            var info = block.FencedCodeData == null ? null : block.FencedCodeData.Info;
+            if (info != null && info.Length > 0)
+            {
+                x = info.IndexOf(' ');
+                if (x == -1)
+                    x = info.Length;
+
+                WriteConstant(" class=\"language-");
+                EscapeHtml(new StringPart(info, 0, x));
+                writer.Write('\"');
+            }
+            Write('>');
+            EscapeHtml(block.StringContent);
+            WriteLineConstant("</code></pre>");
+        }
+
+        protected virtual void PrintHeader(TextWriter writer, Block block, CommonMarkSettings settings)
+        {
+            int x;
+            EnsureLine();
+
+            x = block.HeaderLevel;
+
+            if (_trackPosition)
+            {
+                WriteConstant("<h" + x.ToString(CultureInfo.InvariantCulture));
+                PrintPosition(block);
+                InlinesToHtml(writer, block.InlineContent, settings);
+                WriteLineConstant(x > 0 && x < 7 ? HeaderCloserTags[x - 1] : "</h" + x.ToString(CultureInfo.InvariantCulture) + ">");
+            }
+            else
+            {
+                WriteConstant(x > 0 && x < 7 ? HeaderOpenerTags[x - 1] : "<h" + x.ToString(CultureInfo.InvariantCulture) + ">");
+                InlinesToHtml(writer, block.InlineContent, settings);
+                WriteLineConstant(x > 0 && x < 7 ? HeaderCloserTags[x - 1] : "</h" + x.ToString(CultureInfo.InvariantCulture) + ">");
+            }
+        }
+
+        protected virtual void PrintList(TextWriter writer, Block block)
+        {
+            // make sure a list starts at the beginning of the line:
+            EnsureLine();
+            var data = block.ListData;
+            WriteConstant(data.ListType == ListType.Bullet ? "<ul" : "<ol");
+            if (data.Start != 1)
+            {
+                WriteConstant(" start=\"");
+                WriteConstant(data.Start.ToString(CultureInfo.InvariantCulture));
+                Write('\"');
+            }
+            if (_trackPosition) PrintPosition(block);
+            writer.WriteLine('>');
+
+            _stackLiteral = data.ListType == ListType.Bullet ? "</ul>" : "</ol>";
```

Comment:
Stack is not exposed to the derived class. As such it seems that the way to go would be to use methods `BeginBlock` and `CompleteBlock` - the first would render the opener and the second would be added to the stack.

This would also enable inlines to be processed without explicitly calling `InlineToHtml` but instead they would be processed automatically (if `BeginBlock` returned that the contents have to be processed by the base class).

See `Block.AsEnumerable()` on how such pattern can be implemented.